File: k9mail/src/main/java/com/fsck/k9/notification/NotificationController.java
package com.fsck.k9.notification;


import android.content.Context;

import androidx.annotation.NonNull;
import androidx.core.app.NotificationManagerCompat;

import com.fsck.k9.Account;
import com.fsck.k9.Clock;
import com.fsck.k9.Preferences;
import com.fsck.k9.activity.MessageReference;
import com.fsck.k9.mail.Folder;
import com.fsck.k9.mailstore.LocalMessage;

import java.util.List;

public class NotificationController {
    private final CertificateErrorNotificationController certificateErrorNotificationController;
    private final AuthenticationErrorNotificationController authenticationErrorNotificationController;
    private final SyncNotificationController syncNotificationController;
    private final SendFailedNotificationController sendFailedNotificationController;
    private final NewMailNotificationController newMailNotificationController;

    private final NotificationChannelManager channelUtils;

    public static NotificationController newInstance(Context context) {
        Context appContext = context.getApplicationContext();
        NotificationManagerCompat notificationManager = NotificationManagerCompat.from(appContext);
        return new NotificationController(appContext, notificationManager);
    }

    NotificationController(Context context, NotificationManagerCompat notificationManager) {
        this.channelUtils = new NotificationChannelManager(context, Preferences.getPreferences(context));
        NotificationResourceProvider notificationResourceProvider = new PlanckNotificationResourceProvider(context);
        NotificationHelper notificationHelper = new NotificationHelper(context, notificationManager, channelUtils, notificationResourceProvider);
        NotificationActionCreator actionBuilder = new NotificationActionCreator(context);
        certificateErrorNotificationController = new CertificateErrorNotificationController(notificationHelper, actionBuilder, notificationResourceProvider);
        authenticationErrorNotificationController = new AuthenticationErrorNotificationController(notificationHelper, actionBuilder, notificationResourceProvider);
        syncNotificationController = new SyncNotificationController(notificationHelper, actionBuilder, notificationResourceProvider);
        sendFailedNotificationController = new SendFailedNotificationController(notificationHelper, actionBuilder, notificationResourceProvider);
        newMailNotificationController = initializeNewMailNotificationController(
                context,
                notificationManager,
                notificationResourceProvider,
                notificationHelper,
                actionBuilder
        );
    }

    @NonNull
    private NewMailNotificationController initializeNewMailNotificationController(
            Context context,
            NotificationManagerCompat notificationManager,
            NotificationResourceProvider notificationResourceProvider,
            NotificationHelper notificationHelper,
            NotificationActionCreator actionBuilder
    ) {
        NotificationContentCreator notificationContentCreator = new NotificationContentCreator(context, notificationResourceProvider);
        NotificationDataStore notificationDataStore = new NotificationDataStore();
        NotificationRepository notificationRepository = new NotificationRepository(notificationDataStore);
        BaseNotificationDataCreator baseNotificationDataCreator = new BaseNotificationDataCreator();
        SingleMessageNotificationDataCreator singleMessageNotificationDataCreator = new SingleMessageNotificationDataCreator();
        SummaryNotificationDataCreator summaryNotificationDataCreator = new SummaryNotificationDataCreator(singleMessageNotificationDataCreator);
        NewMailNotificationManager newMailNotificationManager = new NewMailNotificationManager(
                notificationContentCreator,
                notificationRepository,
                baseNotificationDataCreator,
                singleMessageNotificationDataCreator,
                summaryNotificationDataCreator,
                Clock.INSTANCE
        );
        LockScreenNotificationCreator lockScreenNotificationCreator = new LockScreenNotificationCreator(
                notificationHelper, notificationResourceProvider
        );
        SingleMessageNotificationCreator singleMessageNotificationCreator = new SingleMessageNotificationCreator(
                notificationHelper, actionBuilder, notificationResourceProvider, lockScreenNotificationCreator
        );
        SummaryNotificationCreator summaryNotificationCreator = new SummaryNotificationCreator(
                notificationHelper, actionBuilder, lockScreenNotificationCreator, singleMessageNotificationCreator, notificationResourceProvider
        );
        return new NewMailNotificationController(
                notificationManager,
                newMailNotificationManager,
                summaryNotificationCreator,
                singleMessageNotificationCreator
        );
    }

    public void showCertificateErrorNotification(Account account, boolean incoming) {
        certificateErrorNotificationController.showCertificateErrorNotification(account, incoming);
    }

    public void clearCertificateErrorNotifications(Account account, boolean incoming) {
        certificateErrorNotificationController.clearCertificateErrorNotifications(account, incoming);
    }

    public void showAuthenticationErrorNotification(Account account, boolean incoming) {
        authenticationErrorNotificationController.showAuthenticationErrorNotification(account, incoming);
    }

    public void clearAuthenticationErrorNotification(Account account, boolean incoming) {
        authenticationErrorNotificationController.clearAuthenticationErrorNotification(account, incoming);
    }

    public void showSendingNotification(Account account) {
        syncNotificationController.showSendingNotification(account);
    }

    public void clearSendingNotification(Account account) {
        syncNotificationController.clearSendingNotification(account);
    }

    public void showSendFailedNotification(Account account, Exception exception) {
        sendFailedNotificationController.showSendFailedNotification(account, exception);
    }

    public void clearSendFailedNotification(Account account) {
        sendFailedNotificationController.clearSendFailedNotification(account);
    }

    public void showFetchingMailNotification(Account account, Folder folder) {
        syncNotificationController.showFetchingMailNotification(account, folder);
    }

    public void clearFetchingMailNotification(Account account) {
        syncNotificationController.clearFetchingMailNotification(account);
    }

    public void addNewMailsNotification(Account account, List<LocalMessage> messages, int previousUnreadMessageCount) {
        newMailNotificationController.addNewMailsNotification(account, messages);
    }

    public void removeNewMailNotification(Account account, MessageReference messageReference) {
        newMailNotificationController.removeNewMailNotification(account, messageReference);
    }

    public void clearNewMailNotifications(Account account) {
        newMailNotificationController.clearNewMailNotifications(account);
    }

    public void clearNewMailNotifications(Account account, String folderName) {
        newMailNotificationController.clearNewMailNotifications(account, folderName);
    }

    public void updateChannels() {
        channelUtils.updateChannels();
    }
}

File: k9mail/src/main/java/com/fsck/k9/notification/NotificationController.kt
package com.fsck.k9.notification

import android.content.Context
import androidx.core.app.NotificationManagerCompat
import com.fsck.k9.Account
import com.fsck.k9.Clock
import com.fsck.k9.Preferences
import com.fsck.k9.RealClock
import com.fsck.k9.activity.MessageReference
import com.fsck.k9.mail.Folder
import com.fsck.k9.mailstore.LocalMessage
import security.planck.notification.GroupMailInvite
import security.planck.notification.GroupMailSignal

class NotificationController internal constructor(
    context: Context,
    notificationManager: NotificationManagerCompat,
    private val clock: Clock = RealClock()
) {
    private val certificateErrorNotificationController: CertificateErrorNotificationController
    private val authenticationErrorNotificationController: AuthenticationErrorNotificationController
    private val syncNotificationController: SyncNotificationController
    private val sendFailedNotificationController: SendFailedNotificationController
    private val groupedNotificationController: GroupedNotificationController
    private val channelUtils: NotificationChannelManager

    init {
        channelUtils = NotificationChannelManager(context, Preferences.getPreferences(context))
        val notificationResourceProvider: NotificationResourceProvider =
            PlanckNotificationResourceProvider(context)
        val notificationHelper = NotificationHelper(
            context,
            notificationManager,
            channelUtils,
            notificationResourceProvider
        )
        val actionBuilder = NotificationActionCreator(context)
        certificateErrorNotificationController = CertificateErrorNotificationController(
            notificationHelper,
            actionBuilder,
            notificationResourceProvider
        )
        authenticationErrorNotificationController = AuthenticationErrorNotificationController(
            notificationHelper,
            actionBuilder,
            notificationResourceProvider
        )
        syncNotificationController = SyncNotificationController(
            notificationHelper,
            actionBuilder,
            notificationResourceProvider
        )
        sendFailedNotificationController = SendFailedNotificationController(
            notificationHelper,
            actionBuilder,
            notificationResourceProvider
        )
        groupedNotificationController = initializeGroupedNotificationController(
            context,
            notificationResourceProvider,
            notificationHelper,
            actionBuilder
        )
    }

    private fun initializeGroupedNotificationController(
        context: Context,
        notificationResourceProvider: NotificationResourceProvider,
        notificationHelper: NotificationHelper,
        actionBuilder: NotificationActionCreator
    ): GroupedNotificationController {
        val singleMessageNotificationDataCreator = SingleGroupedNotificationDataCreator()
        val summaryNotificationDataCreator =
            SummaryGroupedNotificationDataCreator(singleMessageNotificationDataCreator)
        val groupedNotificationManager  =
            GroupedNotificationManager(
                contentCreator = NotificationContentCreator(context, notificationResourceProvider),
                newMailNotificationRepository = NotificationRepository(NotificationDataStore(NotificationGroupType.NEW_MAIL)),
                groupMailNotificationRepository = NotificationRepository(NotificationDataStore(NotificationGroupType.GROUP_MAIL)),
                baseNotificationDataCreator = BaseNotificationDataCreator(),
                singleMessageNotificationDataCreator = singleMessageNotificationDataCreator,
                summaryNotificationDataCreator = summaryNotificationDataCreator,
                clock = clock
            )
        val newMailScreenNotificationCreator = NewMailLockScreenNotificationCreator(
            notificationHelper, notificationResourceProvider
        )
        val groupMailScreenNotificationCreator = GroupMailLockScreenNotificationCreator(
            notificationHelper, notificationResourceProvider
        )
        val singleNewMailNotificationCreator = SingleMessageNotificationCreator(
                notificationHelper,
                actionBuilder,
                notificationResourceProvider,
                newMailScreenNotificationCreator
            )
        val singleGroupMailNotificationCreator = SingleGroupMailNotificationCreator(
            notificationHelper,
            actionBuilder,
            notificationResourceProvider,
            groupMailScreenNotificationCreator
        )
        val summaryNewMailNotificationCreator = SummaryNewMailNotificationCreator(
            notificationHelper,
            actionBuilder,
            newMailScreenNotificationCreator,
            singleNewMailNotificationCreator,
            notificationResourceProvider
        )
        val summaryGroupMailNotificationCreator = SummaryGroupMailNotificationCreator(
            notificationHelper,
            actionBuilder,
            groupMailScreenNotificationCreator,
            singleGroupMailNotificationCreator,
            notificationResourceProvider
        )
        return GroupedNotificationController(
            notificationHelper = notificationHelper,
            groupedNotificationManager = groupedNotificationManager,
            newMailSummaryNotificationCreator = summaryNewMailNotificationCreator,
            newMailSingleNotificationCreator = singleNewMailNotificationCreator,
            groupMailSingleNotificationCreator = singleGroupMailNotificationCreator,
            groupMailSummaryNotificationCreator = summaryGroupMailNotificationCreator
        )
    }

    fun showCertificateErrorNotification(account: Account?, incoming: Boolean) {
        certificateErrorNotificationController.showCertificateErrorNotification(account!!, incoming)
    }

    fun clearCertificateErrorNotifications(account: Account?, incoming: Boolean) {
        certificateErrorNotificationController.clearCertificateErrorNotifications(
            account!!,
            incoming
        )
    }

    fun showAuthenticationErrorNotification(account: Account?, incoming: Boolean) {
        authenticationErrorNotificationController.showAuthenticationErrorNotification(
            account!!,
            incoming
        )
    }

    fun clearAuthenticationErrorNotification(account: Account?, incoming: Boolean) {
        authenticationErrorNotificationController.clearAuthenticationErrorNotification(
            account!!,
            incoming
        )
    }

    fun showSendingNotification(account: Account?) {
        syncNotificationController.showSendingNotification(account!!)
    }

    fun clearSendingNotification(account: Account?) {
        syncNotificationController.clearSendingNotification(account!!)
    }

    fun showSendFailedNotification(account: Account?, exception: Exception?) {
        sendFailedNotificationController.showSendFailedNotification(account!!, exception!!)
    }

    fun clearSendFailedNotification(account: Account?) {
        sendFailedNotificationController.clearSendFailedNotification(account!!)
    }

    fun showFetchingMailNotification(account: Account?, folder: Folder<*>?) {
        syncNotificationController.showFetchingMailNotification(account!!, folder!!)
    }

    fun clearFetchingMailNotification(account: Account?) {
        syncNotificationController.clearFetchingMailNotification(account!!)
    }

    fun addNewMailsNotification(
        account: Account,
        messages: List<LocalMessage>,
        previousUnreadMessageCount: Int
    ) {
        groupedNotificationController.addNewMailsNotification(account, messages)
    }

    fun clearNewMailNotifications(account: Account, folderName: String) {
        groupedNotificationController.removeNewMailNotifications(account) {
            it.filter { messageReference -> messageReference.folderName == folderName }
        }
    }

    fun removeNewMailNotification(account: Account, messageReference: MessageReference) {
        groupedNotificationController.removeNewMailNotifications(account) {
            it.filter { reference -> reference == messageReference }
        }
    }

    fun clearGroupedNotifications(account: Account) {
        groupedNotificationController.clearNewMailNotifications(account)
        groupedNotificationController.clearGroupMailNotifications(account)
    }

    fun clearNewMailNotifications(account: Account) {
        groupedNotificationController.clearNewMailNotifications(account)
    }

    fun addGroupMailNotification(account: Account, groupMailSignal: GroupMailSignal) {
        groupedNotificationController.addGroupMailNotification(
            account,
            groupMailSignal,
            false
        )
    }

    fun removeGroupMailNotification(account: Account, groupMailInvite: GroupMailInvite) {
        groupedNotificationController.removeGroupMailNotifications(account) {
            it.filter { reference -> reference == groupMailInvite }
        }
    }

    fun clearGroupMailNotifications(account: Account) {
        groupedNotificationController.clearGroupMailNotifications(account)
    }

    fun updateChannels() {
        channelUtils.updateChannels()
    }

    companion object {
        @JvmStatic
        fun newInstance(context: Context): NotificationController {
            val appContext = context.applicationContext
            val notificationManager = NotificationManagerCompat.from(appContext)
            return NotificationController(appContext, notificationManager)
        }
    }
}